geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexKeySpace.scala

```diff
@@ -59,4 +63,99 @@ trait IndexKeySpace[T] {
     * @return
     */
   def getIndexValues(sft: SimpleFeatureType, filter: Filter, explain: Explainer): T
+
+  /**
+    * Creates ranges over the index keys
+    *
+    * @param values index values @see getIndexValues
+    * @return
+    */
+  def getRanges(values: T): Iterator[ScanRange[U]]
+
+  /**
+    * Creates bytes from ranges
+    *
+    * @param ranges typed scan ranges. @see `getRanges`
+    * @param prefixes prefixes to the range, if any. each prefix will create a new range
+    * @return
+    */
+  def getRangeBytes(ranges: Iterator[ScanRange[U]],
+                    prefixes: Seq[Array[Byte]] = Seq.empty,
+                    tier: Boolean = false): Iterator[ByteRange]
+
+  /**
+    * Determines if the ranges generated by `getRanges` are sufficient to fulfill the query,
+    * or if additional filtering needs to be done
+    *
+    * @param config data store config
+    * @param values index values @see getIndexValues
+    * @param hints query hints
+    * @return
+    */
+  def useFullFilter(values: Option[T],
+                    config: Option[GeoMesaDataStoreConfig],
+                    hints: Hints): Boolean
+}
+
+object IndexKeySpace {
+
+  type ToIndexKeyBytes = (Seq[Array[Byte]], SimpleFeature, Array[Byte]) => Seq[Array[Byte]]
+
+  sealed trait ByteRange
+
+  case class BoundedByteRange(lower: Array[Byte], upper: Array[Byte]) extends ByteRange
+  case class SingleRowByteRange(row: Array[Byte]) extends ByteRange
+  case class TieredByteRange(lower: Array[Byte],
```

Comment:
Ok, how's a TieredByteRange different from a BoundedByteRange?

Do you just need to pass around the pair of Booleans to indicated whether or not key ranges can be prepended and/or appended?

Reply:
yeah - the bounded byte range is assumed to be both prependable and appendable, otherwise you'd use this one. I'll add some more scaladocs... :)